Add a functiton to create or remove ovs bridge

Yingshun opened the pull request:
In some cases, it needs to reobtain IP address after adding port,
so add a new function to create/delete ovs bridge via tmux command
on local and remote host.

Signed-off-by: Yingshun Cui <yicui@redhat.com>

PaulYuuu commented:
What does "network connection breaks if adding ovs port" mean?
And "openvswitch2.11" could not cover all versions of ovs, let users choose themselves.
```
openvswitch2.11-2.11.3-67.el8fdp.ppc64le : Open vSwitch
Repo        : FDP
Matched from:
Filename    : /usr/bin/ovs-vsctl

openvswitch2.12-2.12.0-23.el8fdp.ppc64le : Open vSwitch
Repo        : FDP
Matched from:
Filename    : /usr/bin/ovs-vsctl

openvswitch2.13-2.13.0-58.el8fdp.ppc64le : Open vSwitch
Repo        : FDP
Matched from:
Filename    : /usr/bin/ovs-vsctl
```

Yingshun commented:
> What does "network connection breaks if adding ovs port" mean?
> And "openvswitch2.11" could not cover all versions of ovs, let users choose themselves.
> 
> ```
> openvswitch2.11-2.11.3-67.el8fdp.ppc64le : Open vSwitch
> Repo        : FDP
> Matched from:
> Filename    : /usr/bin/ovs-vsctl
> 
> openvswitch2.12-2.12.0-23.el8fdp.ppc64le : Open vSwitch
> Repo        : FDP
> Matched from:
> Filename    : /usr/bin/ovs-vsctl
> 
> openvswitch2.13-2.13.0-58.el8fdp.ppc64le : Open vSwitch
> Repo        : FDP
> Matched from:
> Filename    : /usr/bin/ovs-vsctl
> ```

Thanks for your comment. I've updated the commit message. If we add the only linked network device to bridge, we need to release and re-obtain IP address, so I use tmux command instead.
And about the package, I've change it to "openvswitch2.*", is it ok to you? In libvirt-ci, we don't have 2.12 and 2.13, so If you have other ideas about how to install the package, could you please provide some details?

PaulYuuu commented:
@Yingshun, I suggest removing the installation step in this patch. To invoke this method, users need to ensure the host/guest has already the ovs environment so that they can tweak the ovs version to be tested.

`nmcli` can also create the ovs bridge without reobtain IP address, but it isn't the public interface, just let you know.
```
dnf install -y NetworkManager-ovs
systemctl restart NetworkManager openvswitch

netdev=$(ip -4 route list default | head -1 | grep -Po '(?<=dev )(\S+)')
conid=$(nmcli -g GENERAL.CONNECTION device show ${netdev})
netmac=$(nmcli -g GENERAL.HWADDR device show ${netdev} | sed -e 's/\\//g')

nmcli connection add type ovs-bridge ifname switch con-name switch
nmcli connection add type ovs-port ifname switch master switch con-name ovs-port-switch
nmcli connection add type ovs-interface slave-type ovs-port ifname switch master ovs-port-switch con-name ovs-if-switch ipv4.method auto 802-3-ethernet.cloned-mac-address ${netmac} connection.autoconnect yes

nmcli connection add type ovs-port ifname ${netdev} master switch con-name ovs-port-${netdev}
nmcli connection modify "${conid}" master ovs-port-${netdev}
nmcli connection up "${conid}"
```

In summary, better to split it with 2 methods(create/delete), take care that the IP address will be changed from DHCP if bridge mac != interface mac and you will lose your session connection.
Since it is not called too frequently, you can also consider putting it in `tp-libvirt/provider`.

Yingshun commented:
> @Yingshun, I suggest removing the installation step in this patch. To invoke this method, users need to ensure the host/guest has already the ovs environment so that they can tweak the ovs version to be tested.
> 
> `nmcli` can also create the ovs bridge without reobtain IP address, but it isn't the public interface, just let you know.
> 
> ```
> dnf install -y NetworkManager-ovs
> systemctl restart NetworkManager openvswitch
> 
> netdev=$(ip -4 route list default | head -1 | grep -Po '(?<=dev )(\S+)')
> conid=$(nmcli -g GENERAL.CONNECTION device show ${netdev})
> netmac=$(nmcli -g GENERAL.HWADDR device show ${netdev} | sed -e 's/\\//g')
> 
> nmcli connection add type ovs-bridge ifname switch con-name switch
> nmcli connection add type ovs-port ifname switch master switch con-name ovs-port-switch
> nmcli connection add type ovs-interface slave-type ovs-port ifname switch master ovs-port-switch con-name ovs-if-switch ipv4.method auto 802-3-ethernet.cloned-mac-address ${netmac} connection.autoconnect yes
> 
> nmcli connection add type ovs-port ifname ${netdev} master switch con-name ovs-port-${netdev}
> nmcli connection modify "${conid}" master ovs-port-${netdev}
> nmcli connection up "${conid}"
> ```
> 
> In summary, better to split it with 2 methods(create/delete), take care that the IP address will be changed from DHCP if bridge mac != interface mac and you will lose your session connection.
> Since it is not called too frequently, you can also consider putting it in `tp-libvirt/provider`.

Thanks for your comments and tips, both of them are very useful. I've removed openvswitch pkg installation part and split it to 2 functions. And they will be used in migration tests, so I'd like to add them into avocado-vt.

Yingshun commented:
**Test result:**
```
# avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_network.ovs_br.without_postcopy 
JOB ID     : 9d1bc1b5f3ef0c54922a0f515013186704822e1b
JOB LOG    : /root/avocado/job-results/job-2020-09-10T02.28-9d1bc1b/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_network.ovs_br.without_postcopy: PASS (317.07 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

```

PaulYuuu commented:
Modify the public network during the test, this operation is very dangerous. So let's wait for the opinions of other maintainers.

Yingshun commented on virttest/utils_net.py:
```diff
@@ -4247,3 +4247,49 @@ def get_netkvm_param_value(vm, param):
+    if not utils_package.package_install("tmux", session):
+        raise exceptions.TestError("Failed to install the tmux packages.")
+
+    cmd = "ovs-vsctl add-br {0};ovs-vsctl add-port {0} {1};dhclient -r;"\
```
This tool is only used by this function for now, but I will think about your suggestion, Thanks!

PaulYuuu replied:
Maybe I didn’t explain it clearly. I mean do you want to use `tmux -c` with `ovs-vsctl` like `tmux -c "command1;command2"`?

Yingshun replied:
Ohh, yes, missed that, will fix it

Yingshun replied:
Updated.

Yingshun commented:
> Modify the public network during the test, this operation is very dangerous. So let's wait for the opinions of other maintainers.

Yes, I agree. @kylazhang @dzhengfy  would you please help review it?

kylazhang commented on virttest/utils_net.py:
```diff
@@ -4247,3 +4247,51 @@ def get_netkvm_param_value(vm, param):
+        runner = session.cmd
+    iface_name = get_net_if(runner=runner, state="UP")[0]
+    if not utils_package.package_install("tmux", session):
+        raise exceptions.TestError("Failed to install the tmux packages.")
```
we may also need to check the openvswitch package install status, please also check the func setup_ovs_vhostuser and in it there are checks, I think maybe you can reuse or make them a separate func to avoid the code duplication

Yingshun replied:
Thanks @kylazhang ! Updated.

Yingshun commented:
> Modify the public network during the test, this operation is very dangerous. So let's wait for the opinions of other maintainers.

We have some cases which modify the public network in tp-libvirt, they are iface_unprivileged, virtual_network_multivms and iface_bridge, and still have utils_sriov in avocado-vt. They are implemented using 'tmux' command line. It looks like they are stable for now.

kylazhang commented on virttest/utils_net.py:
```diff
@@ -4247,3 +4247,61 @@ def get_netkvm_param_value(vm, param):
+    if not utils_package.package_install("tmux", session):
+        raise exceptions.TestError("Failed to install the tmux packages.")
+
+    res = utils_misc.cmd_status_output("which ovs-vsctl", shell=True,
```
please use path.find_command

kylazhang replied:
Ok, sorry forget to think about the remote checking issue

Yingshun replied:
It's Ok.

kylazhang commented on virttest/utils_net.py:
```diff
@@ -4247,3 +4247,61 @@ def get_netkvm_param_value(vm, param):
+    res = utils_misc.cmd_status_output("which ovs-vsctl", shell=True,
+                                       ignore_status=False, session=session)[0]
+    if res == 1:
+        logging.error("ovs-vsctl: command not found, please make sure the "
```
I think raise exception error should be better

Yingshun replied:
Yes, done.

kylazhang commented on virttest/utils_net.py:
```diff
@@ -4247,3 +4247,61 @@ def get_netkvm_param_value(vm, param):
+                                       ignore_status=False, session=session)[0]
+    if res == 1:
+        logging.error("ovs-vsctl: command not found, please make sure the "
+                      "openvswitch2 pkg is installed.")
```
openvswitch/openvswitch2

Yingshun replied:
Updated.

dzhengfy reviewed (approved):
LGTM

dzhengfy commented:
As PaulYuuu's comments have been answered, and this PR has been over 1 month, I merge it now.
